daffodil-test/src/test/scala/org/apache/daffodil/layers/IPv4Checksum.scala

```diff
@@ -0,0 +1,239 @@
+    //
+    // For parsing, all the hard work happens here, allowing the layered input stream to
+    // just deliver the bytes
+    //
+    Assert.invariant(s.bitPos0b % 8 == 0) // we are byte aligned.
+    Assert.invariant(optOriginalInputStream.isDefined)
+    val bitLen = hdrLengthInBytes * 8
+
+    try
+      IOUtils.readFully(optOriginalInputStream.get, hdrByteArr)
+    catch {
+      case eof: EOFException =>
+        throw DataInputStream.NotEnoughDataException(bitLen)
+    }
+    val checksumInDataStream = shortBuf.get()
+    val checksum = computeChecksum(shortBuf)
+    s.setVariable(finalChecksumVRD, checksum, srd) // assign to result variable.
+  }
+
+  final class SuspendableChecksumLayerOperation()
+    extends SuspendableOperation {
+    override def rd = srd
+
+    /**
+     * Test succeeds if all the data required has been written to the layer
+     */
+    protected def test(ustate: UState) = {
+      limitingOutputStream.size() == hdrLengthInBytes
+    }
+
+    /**
+     * Computes checksum and overwrites that part of the layer data
+     * with the new checksum, writes the layer data out,
+     * and assigns the output variable.
+     */
+    protected def continuation(ustate: UState): Unit = {
+      Assert.invariant(optOriginalOutputStream.isDefined)
+      val finalChecksum = computeChecksum(shortBuf)
+      ustate.setVariable(finalChecksumVRD, finalChecksum, srd) // assign to the result variable.
+      //
+      // clobber the byte buffer bytes corresponding to the checksum with
+      // the recomputed value
+      //
+      shortBuf.put(chksumShortIndex, finalChecksum.toShort)
+      //
+      // write out the layer data (which has recomputed checksum in it.
+      optOriginalOutputStream.get.write(hdrByteArr)
+      optOriginalOutputStream.get.close()
+    }
+  }
+
+  private lazy val suspendableOperation = new SuspendableChecksumLayerOperation()
+
+  override def endLayerForUnparse(s: UState): Unit = {
+    suspendableOperation.run(s)
+  }
+
+}
+
+  object IPv4Checksum
+    extends LayerTransformerFactory("IPv4Checksum") {
+
+    override def newInstance(
+      maybeLayerCharsetEv: Maybe[LayerCharsetEv],
+      maybeLayerLengthKind: Maybe[LayerLengthKind],
+      maybeLayerLengthEv: Maybe[LayerLengthEv],
+      maybeLayerLengthUnits: Maybe[LayerLengthUnits],
+      maybeLayerBoundaryMarkEv: Maybe[LayerBoundaryMarkEv],
+      srd: SequenceRuntimeData) = {
+      srd.schemaDefinitionUnless(maybeLayerLengthKind.isEmpty ||
+        maybeLayerLengthKind.get == LayerLengthKind.Implicit,
+        "Must have dfdlx:layerLengthKind undefined or 'implicit', but was '%s'.",
+        maybeLayerLengthKind.get.toString)
+      //
+      // ToDo: We can't issue SDW because we don't have a context object
+      // for that sort of compilation time warning. We *should* have that.
+      // That requires a layer factory method called at schema compile time with the
+      // Sequence term (not just term runtime data) as an argument.
```

Comment:
Is this comment out of date?  I see SequenceRuntimeData being passed above, so it looks like you've already made that change.  If so, please delete the above comment.

Reply:
An SRD isn't enough to issue schema def warnings. Warnings are accumulated at compile time, so have to be issued on a compiler object such as the LocalSequence object. An SRD is enough to issue a runtime SDE, it would be good if we got some of these also at schema compile time. 

I'm leaving fixing this off to the other ticket to refactor these layers to they are pluggable.

Reply:
Actually, I take it back. You can issue schema definition warnings on a runtime data object. But we really want compilation time errors, not runtime errors.